tensor2tensor/layers/bayes.py

```diff
@@ -41,6 +42,50 @@ def softplus():  # alias, following tf.keras.constraints
   return Softplus()
 
 
+class Zeros(object):
+  """Function returning zeros tensor of same shape excluding the last dim."""
+
+  def __call__(self, inputs):
+    return tf.zeros(tf.shape(inputs)[:-1], inputs.dtype)
+
+  def get_config(self):
+    return {}
+
+
+class ExponentiatedQuadratic(object):
+  """Exponentiated quadratic kernel."""
+
+  def __init__(self, variance, lengthscale):
+    self.variance = variance
+    self.lengthscale = lengthscale
+
+  def __call__(self, x1, x2):
+    """Computes exponentiated quadratic over all pairs of inputs.
+
+    Args:
+      x1: Tensor of shape [batch_x1, ...]. Slices along the batch axis denote an
+        individual input to be passed to the kernel. It is computed pairwise
+        with each input sliced from x2.
+      x2: Tensor of shape [batch_x2, ...]. Slices along the batch axis denote an
+        individual input passed to the kernel function. It is computed pairwise
+        with each input sliced from x1.
+
+    Returns:
+      Tensor of shape [batch_x1, batch_x2].
+    """
+    x1 = x1 / self.lengthscale
+    x2 = x2 / self.lengthscale
+    x1_squared = tf.reduce_sum(tf.square(x1), list(range(1, len(x1.shape))))
+    x2_squared = tf.reduce_sum(tf.square(x2), list(range(1, len(x2.shape))))
+    square = (tf.reshape(x1_squared, [-1, 1]) +
+              tf.reshape(x2_squared, [1, -1]) -
+              2 * tf.matmul(x1, x2, transpose_b=True))
```

Comment:
It seems like the docs and part of this function aim to have a kernel that supports input tensors with dims > 2, but this matmul won't work for that case, right?  Imagine x1 is shape [batch_size_x1, dim1, dim2] and x2 is shape [batch_size_x2, dim1, dim2].  This would transpose the latter two axes of x2, and attempt an element-wise multiplication across the batch axis, but that would fail unless batch_size_x1 == batch_size_x2.  We could use a tensor dot product here, such as `tf.tensordot(x1, tf.transpose(x2), axes=len(x1.shape)-1)`, which would yield an output with shape `[batch_size_x1, batch_size_x2]` regardless of the number of tensor dimensions per example.

Reply:
Good pt. For now, raised NotImplementedError for dims > 2.